alpaka math functions does not work in lambdas with alpaka 0.9 and CUDA backend

# Problem

I want to pass a lambda to a kernel, which is called and use a `alpaka::math` function. Therefore the lambda needs an `acc` object as argument. The CUDA backend does not support generic lambdas, therefore it needs to be an explicit type. This causes a compiler error. 

```bash
FAILED: CMakeFiles/alpaka_lambda.dir/main.cpp.o 
/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/g++ -DALPAKA_ACC_GPU_CUDA_ENABLED -DALPAKA_BLOCK_SHARED_DYN_MEMBER_ALLOC_KIB=47 -DALPAKA_DEBUG=0 -DALPAKA_DEBUG_OFFLOAD_ASSUME_HOST -DALPAKA_OFFLOAD_MAX_BLOCK_SIZE=256 -DBOOST_ALL_NO_LIB -isystem=/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/targets/x86_64-linux/include -isystem=/home/sehrig/application/miniconda3/envs/vikunja-dev/include -isystem=/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/boost-1.74.0-lqyr7ou7iafiwvq5fb6ipp62o6l2gqlf/include -isystem=/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/include --generate-code=arch=compute_52,code=[compute_52,sm_52] -Xcompiler=-fPIE --extended-lambda --expt-relaxed-constexpr -Xcudafe=--display_error_number -Xcudafe=--diag_suppress=esa_on_defaulted_function_ignored -Xcompiler -pthread -std=c++17 -MD -MT CMakeFiles/alpaka_lambda.dir/main.cpp.o -MF CMakeFiles/alpaka_lambda.dir/main.cpp.o.d -x cu -c ../main.cpp -o CMakeFiles/alpaka_lambda.dir/main.cpp.o
/home/sehrig/application/miniconda3/envs/vikunja-dev/include/alpaka/math/Traits.hpp(850): error #3050: calling a __device__ function("operator()") from a __host__ __device__ function("max") is not allowed
          detected during instantiation of "auto alpaka::math::max(const T &, const Tx &, const Ty &) [with T=alpaka::AccGpuCudaRt<alpaka::DimInt<1UL>, std::size_t>, Tx=float, Ty=float]" 
../main.cpp(40): here

/home/sehrig/application/miniconda3/envs/vikunja-dev/include/alpaka/math/Traits.hpp(850): error #3050: calling a __device__ function("operator()") from a __host__ __device__ function("max") is not allowed
          detected during instantiation of "auto alpaka::math::max(const T &, const Tx &, const Ty &) [with T=alpaka::AccGpuCudaRt<alpaka::DimInt<1UL>, std::size_t>, Tx=float, Ty=float]" 
../main.cpp(40): here

2 errors detected in the compilation of "../main.cpp".
ninja: build stopped: subcommand failed.
```

The same code is working with the serial backend, OpenMP backend and also the Hip backend. Maybe it also happen with specific CUDA versions. CUDA 11.0 failed and CUDA 11.2 was fine.

# Code to reproduce

Use the alpaka version of commit [ebc1171](https://github.com/alpaka-group/alpaka/commit/ebc1171feac21f1e21c49bcd9f053e7b01b584d0) (branch `0.9.0-rc1`).

CMakeLists.txt
```cmake
cmake_minimum_required(VERSION 3.18)

project(alpaka_lambda)
set(CMAKE_POSITION_INDEPENDENT_CODE ON)
set(alpaka_CUDA_EXPT_EXTENDED_LAMBDA ON)

option(alpaka_ACC_GPU_CUDA_ENABLE "enable alpaka CUDA accelerator" ON)

find_package(alpaka)

alpaka_add_executable(${PROJECT_NAME} main.cpp)
target_link_libraries(${PROJECT_NAME} PRIVATE alpaka::alpaka)
```

main.cpp
```c++
#include <alpaka/alpaka.hpp>


struct GlobalKernel
{
    template<typename TAcc, typename TFunctor>
    ALPAKA_FN_ACC auto operator()(TAcc const& acc, TFunctor const& functor) const -> void
    {
        float f = functor(acc, 42.f);
        printf("%f\n", f);
    }
};

int main()
{
    using Dim = alpaka::DimInt<1>;
    using Idx = std::size_t;

    using Acc = alpaka::AccGpuCudaRt<Dim, Idx>;
    using Queue = alpaka::Queue<Acc, alpaka::Blocking>;

    auto const devAcc = alpaka::getDevByIdx<Acc>(0u);
    Queue queue(devAcc);

    using Vec = alpaka::Vec<Dim, Idx>;
    Vec const elementsPerThread(Vec::all(static_cast<Idx>(1)));
    Vec const threadsPerGrid(Vec::all(static_cast<Idx>(1)));
    using WorkDiv = alpaka::WorkDivMembers<Dim, Idx>;
    WorkDiv const workDiv = alpaka::getValidWorkDiv<Acc>(
        devAcc,
        threadsPerGrid,
        elementsPerThread,
        false,
        alpaka::GridBlockExtentSubDivRestrictions::Unrestricted);

    // does not work
    auto functor = [] ALPAKA_FN_HOST_ACC(alpaka::AccGpuCudaRt<Dim, Idx> const& acc, float const i) -> float
    {
        // return i + 1.f;
        return alpaka::math::max(acc, 41.f, i) + 0.5f;
    };
    GlobalKernel kernel;

    alpaka::exec<Acc>(queue, workDiv, kernel, functor);
    alpaka::wait(queue);

    return 0;
}
```

# Possible solution

@psychocoderHPC points me, that is required that the CUDA implementation of `max` needs an explicit return type.
I fixed the type of this function to `float` and looks like, that it works. 
https://github.com/alpaka-group/alpaka/blob/b074b0df68a96321dc73261ab2b9d3d41180f18c/include/alpaka/math/MathUniformCudaHipBuiltIn.hpp#L642

--- a/include/alpaka/math/Complex.hpp
+++ b/include/alpaka/math/Complex.hpp
@@ -356,6 +356,9 @@ namespace alpaka
 
     //! Host-only math functions matching std::complex<T>.
     //!
+    //! Due to issue #1688, these functions are technically marked host-device and suppress related warnings.
+    //! However, they must be called for host only.
+    //!
     //! They take and return alpaka::Complex (or a real number when appropriate).
     //! Internally cast, fall back to std::complex implementation and cast back.
     //! These functions can be used directly on the host side.
@@ -368,176 +371,201 @@ namespace alpaka
     //!
 
     //! Absolute value
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST T abs(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC T abs(Complex<T> const& x)
     {
         return std::abs(std::complex<T>(x));
     }
 
     //! Arc cosine
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> acos(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> acos(Complex<T> const& x)
     {
         return std::acos(std::complex<T>(x));
     }
 
     //! Arc hyperbolic cosine
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> acosh(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> acosh(Complex<T> const& x)
     {
         return std::acosh(std::complex<T>(x));
     }
 
     //! Argument
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST T arg(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC T arg(Complex<T> const& x)
     {
         return std::arg(std::complex<T>(x));
     }
 
     //! Arc sine
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> asin(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> asin(Complex<T> const& x)
     {
         return std::asin(std::complex<T>(x));
     }
 
     //! Arc hyperbolic sine
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> asinh(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> asinh(Complex<T> const& x)
     {
         return std::asinh(std::complex<T>(x));
     }
 
     //! Arc tangent
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> atan(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> atan(Complex<T> const& x)
     {
         return std::atan(std::complex<T>(x));
     }
 
     //! Arc hyperbolic tangent
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> atanh(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> atanh(Complex<T> const& x)
     {
         return std::atanh(std::complex<T>(x));
     }
 
     //! Complex conjugate
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> conj(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> conj(Complex<T> const& x)
     {
         return std::conj(std::complex<T>(x));
     }
 
     //! Cosine
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> cos(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> cos(Complex<T> const& x)
     {
         return std::cos(std::complex<T>(x));
     }
 
     //! Hyperbolic cosine
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> cosh(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> cosh(Complex<T> const& x)
     {
         return std::cosh(std::complex<T>(x));
     }
 
     //! Exponential
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> exp(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> exp(Complex<T> const& x)
     {
         return std::exp(std::complex<T>(x));
     }
 
     //! Natural logarithm
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> log(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> log(Complex<T> const& x)
     {
         return std::log(std::complex<T>(x));
     }
 
     //! Base 10 logarithm
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> log10(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> log10(Complex<T> const& x)
     {
         return std::log10(std::complex<T>(x));
     }
 
     //! Squared magnitude
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST T norm(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC T norm(Complex<T> const& x)
     {
         return std::norm(std::complex<T>(x));
     }
 
     //! Get a complex number with given magnitude and phase angle
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> polar(T const& r, T const& theta = T())
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> polar(T const& r, T const& theta = T())
     {
         return std::polar(r, theta);
     }
 
     //! Complex power of a complex number
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T, typename U>
-    constexpr ALPAKA_FN_HOST Complex<T> pow(Complex<T> const& x, Complex<U> const& y)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> pow(Complex<T> const& x, Complex<U> const& y)
     {
         return std::pow(std::complex<T>(x), std::complex<U>(y));
     }
 
     //! Real power of a complex number
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T, typename U>
-    constexpr ALPAKA_FN_HOST Complex<T> pow(Complex<T> const& x, U const& y)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> pow(Complex<T> const& x, U const& y)
     {
         return std::pow(std::complex<T>(x), y);
     }
 
     //! Complex power of a real number
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T, typename U>
-    constexpr ALPAKA_FN_HOST Complex<T> pow(T const& x, Complex<U> const& y)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> pow(T const& x, Complex<U> const& y)
     {
         return std::pow(x, std::complex<U>(y));
     }
 
     //! Projection onto the Riemann sphere
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> proj(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> proj(Complex<T> const& x)
     {
         return std::proj(std::complex<T>(x));
     }
 
     //! Sine
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> sin(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> sin(Complex<T> const& x)
     {
         return std::sin(std::complex<T>(x));
     }
 
     //! Hyperbolic sine
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> sinh(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> sinh(Complex<T> const& x)
     {
         return std::sinh(std::complex<T>(x));
     }
 
     //! Square root
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> sqrt(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> sqrt(Complex<T> const& x)
     {
         return std::sqrt(std::complex<T>(x));
     }
 
     //! Tangent
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> tan(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> tan(Complex<T> const& x)
     {
         return std::tan(std::complex<T>(x));
     }
 
     //! Hyperbolic tangent
+    ALPAKA_NO_HOST_ACC_WARNING
     template<typename T>
-    constexpr ALPAKA_FN_HOST Complex<T> tanh(Complex<T> const& x)
+    constexpr ALPAKA_FN_HOST_ACC Complex<T> tanh(Complex<T> const& x)
     {
         return std::tanh(std::complex<T>(x));
     }

--- a/include/alpaka/math/MathUniformCudaHipBuiltIn.hpp
+++ b/include/alpaka/math/MathUniformCudaHipBuiltIn.hpp
@@ -229,7 +229,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Abs<AbsUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_signed_v<TArg>>>
         {
-            __device__ auto operator()(AbsUniformCudaHipBuiltIn const& /* abs_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(AbsUniformCudaHipBuiltIn const& /* abs_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::fabsf(arg);
@@ -254,7 +254,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
             {
                 return sqrt(ctx, arg.real() * arg.real() + arg.imag() * arg.imag());
             }
@@ -264,7 +264,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Acos<AcosUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(AcosUniformCudaHipBuiltIn const& /* acos_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(AcosUniformCudaHipBuiltIn const& /* acos_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::acosf(arg);
@@ -283,7 +283,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
             {
                 // This holds everywhere, including the branch cuts: acos(z) = -i * ln(z + i * sqrt(1 - z^2))
                 return Complex<T>{0.0, -1.0} * log(ctx, arg + Complex<T>{0.0, 1.0} * sqrt(ctx, T(1.0) - arg * arg));
@@ -296,7 +296,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, TArgument const& argument)
+            __host__ __device__ auto operator()(TCtx const& ctx, TArgument const& argument)
             {
                 // Fall back to atan2 so that boundary cases are resolved consistently
                 return atan2(ctx, TArgument{0.0}, argument);
@@ -309,7 +309,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& argument)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& argument)
             {
                 return atan2(ctx, argument.imag(), argument.real());
             }
@@ -319,7 +319,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Asin<AsinUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(AsinUniformCudaHipBuiltIn const& /* asin_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(AsinUniformCudaHipBuiltIn const& /* asin_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::asinf(arg);
@@ -338,7 +338,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
             {
                 // This holds everywhere, including the branch cuts: asin(z) = i * ln(sqrt(1 - z^2) - i * z)
                 return Complex<T>{0.0, 1.0} * log(ctx, sqrt(ctx, T(1.0) - arg * arg) - Complex<T>{0.0, 1.0} * arg);
@@ -349,7 +349,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Atan<AtanUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(AtanUniformCudaHipBuiltIn const& /* atan_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(AtanUniformCudaHipBuiltIn const& /* atan_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::atanf(arg);
@@ -368,7 +368,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
             {
                 // This holds everywhere, including the branch cuts: atan(z) = -i/2 * ln((i - z) / (i + z))
                 return Complex<T>{0.0, -0.5} * log(ctx, (Complex<T>{0.0, 1.0} - arg) / (Complex<T>{0.0, 1.0} + arg));
@@ -383,7 +383,10 @@ namespace alpaka::math
             Tx,
             std::enable_if_t<std::is_floating_point_v<Ty> && std::is_floating_point_v<Tx>>>
         {
-            __device__ auto operator()(Atan2UniformCudaHipBuiltIn const& /* atan2_ctx */, Ty const& y, Tx const& x)
+            __host__ __device__ auto operator()(
+                Atan2UniformCudaHipBuiltIn const& /* atan2_ctx */,
+                Ty const& y,
+                Tx const& x)
             {
                 if constexpr(is_decayed_v<Ty, float> && is_decayed_v<Tx, float>)
                     return ::atan2f(y, x);
@@ -400,7 +403,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Cbrt<CbrtUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_arithmetic_v<TArg>>>
         {
-            __device__ auto operator()(CbrtUniformCudaHipBuiltIn const& /* cbrt_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(CbrtUniformCudaHipBuiltIn const& /* cbrt_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::cbrtf(arg);
@@ -417,7 +420,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Ceil<CeilUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(CeilUniformCudaHipBuiltIn const& /* ceil_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(CeilUniformCudaHipBuiltIn const& /* ceil_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::ceilf(arg);
@@ -434,7 +437,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Conj<ConjUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(ConjUniformCudaHipBuiltIn const& /* conj_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(ConjUniformCudaHipBuiltIn const& /* conj_ctx */, TArg const& arg)
             {
                 return Complex<TArg>{arg, TArg{0.0}};
             }
@@ -444,7 +447,7 @@ namespace alpaka::math
         template<typename T>
         struct Conj<ConjUniformCudaHipBuiltIn, Complex<T>>
         {
-            __device__ auto operator()(ConjUniformCudaHipBuiltIn const& /* conj_ctx */, Complex<T> const& arg)
+            __host__ __device__ auto operator()(ConjUniformCudaHipBuiltIn const& /* conj_ctx */, Complex<T> const& arg)
             {
                 return Complex<T>{arg.real(), -arg.imag()};
             }
@@ -454,7 +457,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Cos<CosUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(CosUniformCudaHipBuiltIn const& /* cos_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(CosUniformCudaHipBuiltIn const& /* cos_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::cosf(arg);
@@ -473,7 +476,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
             {
                 // cos(z) = 0.5 * (exp(i * z) + exp(-i * z))
                 return T(0.5) * (exp(ctx, Complex<T>{0.0, 1.0} * arg) + exp(ctx, Complex<T>{0.0, -1.0} * arg));
@@ -484,7 +487,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Erf<ErfUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(ErfUniformCudaHipBuiltIn const& /* erf_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(ErfUniformCudaHipBuiltIn const& /* erf_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::erff(arg);
@@ -501,7 +504,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Exp<ExpUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(ExpUniformCudaHipBuiltIn const& /* exp_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(ExpUniformCudaHipBuiltIn const& /* exp_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::expf(arg);
@@ -520,7 +523,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
             {
                 // exp(z) = exp(x + iy) = exp(x) * (cos(y) + i * sin(y))
                 auto re = T{}, im = T{};
@@ -533,7 +536,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Floor<FloorUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(FloorUniformCudaHipBuiltIn const& /* floor_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(FloorUniformCudaHipBuiltIn const& /* floor_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::floorf(arg);
@@ -554,7 +557,10 @@ namespace alpaka::math
             Ty,
             std::enable_if_t<std::is_floating_point_v<Tx> && std::is_floating_point_v<Ty>>>
         {
-            __device__ auto operator()(FmodUniformCudaHipBuiltIn const& /* fmod_ctx */, Tx const& x, Ty const& y)
+            __host__ __device__ auto operator()(
+                FmodUniformCudaHipBuiltIn const& /* fmod_ctx */,
+                Tx const& x,
+                Ty const& y)
             {
                 if constexpr(is_decayed_v<Tx, float> && is_decayed_v<Ty, float>)
                     return ::fmodf(x, y);
@@ -573,7 +579,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Isfinite<IsfiniteUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(IsfiniteUniformCudaHipBuiltIn const& /* ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(IsfiniteUniformCudaHipBuiltIn const& /* ctx */, TArg const& arg)
             {
                 return ::isfinite(arg);
             }
@@ -583,7 +589,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Isinf<IsinfUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(IsinfUniformCudaHipBuiltIn const& /* ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(IsinfUniformCudaHipBuiltIn const& /* ctx */, TArg const& arg)
             {
                 return ::isinf(arg);
             }
@@ -593,7 +599,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Isnan<IsnanUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(IsnanUniformCudaHipBuiltIn const& /* ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(IsnanUniformCudaHipBuiltIn const& /* ctx */, TArg const& arg)
             {
                 return ::isnan(arg);
             }
@@ -603,7 +609,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Log<LogUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(LogUniformCudaHipBuiltIn const& /* log_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(LogUniformCudaHipBuiltIn const& /* log_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::logf(arg);
@@ -622,7 +628,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& argument)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& argument)
             {
                 // Branch cut along the negative real axis (same as for std::complex),
                 // principal value of ln(z) = ln(|z|) + i * arg(z)
@@ -638,7 +644,10 @@ namespace alpaka::math
             Ty,
             std::enable_if_t<std::is_arithmetic_v<Tx> && std::is_arithmetic_v<Ty>>>
         {
-            __device__ auto operator()(MaxUniformCudaHipBuiltIn const& /* max_ctx */, Tx const& x, Ty const& y)
+            __host__ __device__ auto operator()(
+                MaxUniformCudaHipBuiltIn const& /* max_ctx */,
+                Tx const& x,
+                Ty const& y)
             {
                 if constexpr(std::is_integral_v<Tx> && std::is_integral_v<Ty>)
                     return ::max(x, y);
@@ -669,7 +678,10 @@ namespace alpaka::math
             Ty,
             std::enable_if_t<std::is_arithmetic_v<Tx> && std::is_arithmetic_v<Ty>>>
         {
-            __device__ auto operator()(MinUniformCudaHipBuiltIn const& /* min_ctx */, Tx const& x, Ty const& y)
+            __host__ __device__ auto operator()(
+                MinUniformCudaHipBuiltIn const& /* min_ctx */,
+                Tx const& x,
+                Ty const& y)
             {
                 if constexpr(std::is_integral_v<Tx> && std::is_integral_v<Ty>)
                     return ::min(x, y);
@@ -700,7 +712,7 @@ namespace alpaka::math
             TExp,
             std::enable_if_t<std::is_floating_point_v<TBase> && std::is_floating_point_v<TExp>>>
         {
-            __device__ auto operator()(
+            __host__ __device__ auto operator()(
                 PowUniformCudaHipBuiltIn const& /* pow_ctx */,
                 TBase const& base,
                 TExp const& exp)
@@ -724,7 +736,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& base, Complex<U> const& exponent)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& base, Complex<U> const& exponent)
             {
                 // pow(z1, z2) = e^(z2 * log(z1))
                 return exp(ctx, exponent * log(ctx, base));
@@ -739,7 +751,7 @@ namespace alpaka::math
             Ty,
             std::enable_if_t<std::is_floating_point_v<Tx> && std::is_floating_point_v<Ty>>>
         {
-            __device__ auto operator()(
+            __host__ __device__ auto operator()(
                 RemainderUniformCudaHipBuiltIn const& /* remainder_ctx */,
                 Tx const& x,
                 Ty const& y)
@@ -762,7 +774,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Round<RoundUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(RoundUniformCudaHipBuiltIn const& /* round_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(RoundUniformCudaHipBuiltIn const& /* round_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::roundf(arg);
@@ -779,7 +791,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Lround<RoundUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(RoundUniformCudaHipBuiltIn const& /* lround_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(RoundUniformCudaHipBuiltIn const& /* lround_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::lroundf(arg);
@@ -796,7 +808,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Llround<RoundUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(RoundUniformCudaHipBuiltIn const& /* llround_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(RoundUniformCudaHipBuiltIn const& /* llround_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::llroundf(arg);
@@ -815,7 +827,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Rsqrt<RsqrtUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_arithmetic_v<TArg>>>
         {
-            __device__ auto operator()(RsqrtUniformCudaHipBuiltIn const& /* rsqrt_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(RsqrtUniformCudaHipBuiltIn const& /* rsqrt_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::rsqrtf(arg);
@@ -834,7 +846,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
             {
                 return T{1.0} / sqrt(ctx, arg);
             }
@@ -844,7 +856,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Sin<SinUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(SinUniformCudaHipBuiltIn const& /* sin_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(SinUniformCudaHipBuiltIn const& /* sin_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::sinf(arg);
@@ -863,7 +875,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
             {
                 // sin(z) = (exp(i * z) - exp(-i * z)) / 2i
                 return (exp(ctx, Complex<T>{0.0, 1.0} * arg) - exp(ctx, Complex<T>{0.0, -1.0} * arg))
@@ -875,7 +887,7 @@ namespace alpaka::math
         template<typename TArg>
         struct SinCos<SinCosUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(
+            __host__ __device__ auto operator()(
                 SinCosUniformCudaHipBuiltIn const& /* sincos_ctx */,
                 TArg const& arg,
                 TArg& result_sin,
@@ -896,7 +908,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(
+            __host__ __device__ auto operator()(
                 TCtx const& ctx,
                 Complex<T> const& arg,
                 Complex<T>& result_sin,
@@ -911,7 +923,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Sqrt<SqrtUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_arithmetic_v<TArg>>>
         {
-            __device__ auto operator()(SqrtUniformCudaHipBuiltIn const& /* sqrt_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(SqrtUniformCudaHipBuiltIn const& /* sqrt_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::sqrtf(arg);
@@ -930,7 +942,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& argument)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& argument)
             {
                 // Branch cut along the negative real axis (same as for std::complex),
                 // principal value of sqrt(z) = sqrt(|z|) * e^(i * arg(z) / 2)
@@ -945,7 +957,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Tan<TanUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(TanUniformCudaHipBuiltIn const& /* tan_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(TanUniformCudaHipBuiltIn const& /* tan_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::tanf(arg);
@@ -964,7 +976,7 @@ namespace alpaka::math
         {
             //! Take context as original (accelerator) type, since we call other math functions
             template<typename TCtx>
-            __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
+            __host__ __device__ auto operator()(TCtx const& ctx, Complex<T> const& arg)
             {
                 // tan(z) = i * (e^-iz - e^iz) / (e^-iz + e^iz) = i * (1 - e^2iz) / (1 + e^2iz)
                 // Warning: this straightforward implementation can easily result in NaN as 0/0 or inf/inf.
@@ -977,7 +989,7 @@ namespace alpaka::math
         template<typename TArg>
         struct Trunc<TruncUniformCudaHipBuiltIn, TArg, std::enable_if_t<std::is_floating_point_v<TArg>>>
         {
-            __device__ auto operator()(TruncUniformCudaHipBuiltIn const& /* trunc_ctx */, TArg const& arg)
+            __host__ __device__ auto operator()(TruncUniformCudaHipBuiltIn const& /* trunc_ctx */, TArg const& arg)
             {
                 if constexpr(is_decayed_v<TArg, float>)
                     return ::truncf(arg);

--- a/test/unit/math/src/TestTemplate.hpp
+++ b/test/unit/math/src/TestTemplate.hpp
@@ -167,7 +167,7 @@ struct TestTemplate
         if(!isFinite(a) && !isFinite(b))
             return true;
         // For the same reason use relative difference comparison with a large margin
-        auto const scalingFactor = static_cast<T>(std::is_same_v<T, float> ? 1e4 : 1e6);
+        auto const scalingFactor = static_cast<T>(std::is_same_v<T, float> ? 1.1e4 : 1.1e6);
         auto const marginValue = scalingFactor * std::numeric_limits<T>::epsilon();
         return (a.real() == Approx(b.real()).margin(marginValue).epsilon(marginValue))
             && (a.imag() == Approx(b.imag()).margin(marginValue).epsilon(marginValue));

--- a/test/unit/math/src/TestTemplate.hpp
+++ b/test/unit/math/src/TestTemplate.hpp
@@ -53,17 +53,21 @@ struct UnderlyingType<alpaka::Complex<T>>
     using type = T;
 };
 
+//! Base test template for math unit tests
+//! @tparam TAcc Accelerator.
+//! @tparam TFunctor Functor defined in Functor.hpp.
 template<typename TAcc, typename TFunctor>
 struct TestTemplate
 {
-    template<typename TData>
-    auto operator()() -> void
+    //! wrappedFunctor is either a TFunctor{} or TFunctor{} wrapped into a host-device lambda
+    template<typename TData, typename TWrappedFunctor = TFunctor>
+    auto operator()(TWrappedFunctor const& wrappedFunctor = TWrappedFunctor{}) -> void
     {
         std::random_device rd{};
         auto const seed = rd();
         std::cout << "testing"
                   << " acc:" << typeid(TAcc).name() << " data type:" << typeid(TData).name()
-                  << " functor:" << typeid(TFunctor).name() << " seed:" << seed << std::endl;
+                  << " functor:" << typeid(TWrappedFunctor).name() << " seed:" << seed << std::endl;
 
         // SETUP (defines and initialising)
         // DevAcc and DevHost are defined in Buffer.hpp too.
@@ -115,10 +119,11 @@ struct TestTemplate
         args.copyToDevice(queue);
         results.copyToDevice(queue);
 
-        auto const taskKernel
-            = alpaka::createTaskKernel<TAcc>(workDiv, kernel, results.pDevBuffer, functor, args.pDevBuffer);
         // Enqueue the kernel execution task.
+        auto const taskKernel
+            = alpaka::createTaskKernel<TAcc>(workDiv, kernel, results.pDevBuffer, wrappedFunctor, args.pDevBuffer);
         alpaka::enqueue(queue, taskKernel);
+
         // Copy back the results (encapsulated in the buffer class).
         results.copyFromDevice(queue);
         alpaka::wait(queue);

--- a/test/unit/math/src/mathFloat.cpp
+++ b/test/unit/math/src/mathFloat.cpp
@@ -76,5 +76,5 @@ TEMPLATE_LIST_TEST_CASE("mathOpsFloat", "[math] [operator]", TestAccFunctorTuple
     using Acc = std::tuple_element_t<0u, TestType>;
     using Functor = std::tuple_element_t<1u, TestType>;
     auto testTemplate = TestTemplate<Acc, Functor>{};
-    testTemplate.template operator()<double>();
+    testTemplate.template operator()<float>();
 }

--- a/test/unit/math/src/mathLambda.cpp
+++ b/test/unit/math/src/mathLambda.cpp
@@ -1,0 +1,93 @@
+/** Copyright 2022 Sergei Bastrakov
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+// For nvcc these tests require extended lambda.
+// Additionally disable the tests for Visual Studio + CUDA since it causes the following compile error:
+// "On Windows, the enclosing parent function ("C_A_T_C_H_T_E_M_P_L_A_T_E_T_E_S_T_F_U_N_C_...") for an extended
+// __host__
+// __device__ lambda cannot have internal or no linkage"
+#if(!defined(__NVCC__) || (defined(__NVCC__) && defined(__CUDACC_EXTENDED_LAMBDA__) && !BOOST_COMP_MSVC))
+
+#    include "Functor.hpp"
+#    include "TestTemplate.hpp"
+
+#    include <alpaka/core/BoostPredef.hpp>
+#    include <alpaka/math/Complex.hpp>
+#    include <alpaka/test/KernelExecutionFixture.hpp>
+#    include <alpaka/test/acc/TestAccs.hpp>
+
+#    include <catch2/catch.hpp>
+
+#    include <cstdint>
+#    include <tuple>
+
+using TestAccs = alpaka::test::EnabledAccs<alpaka::DimInt<1u>, std::size_t>;
+
+// This file only has unit tests for real numbers in order to split the tests between object files
+using UnaryFunctorsReal = alpaka::test::unit::math::UnaryFunctorsReal;
+using BinaryFunctorsReal = alpaka::test::unit::math::BinaryFunctorsReal;
+using UnaryFunctorsComplex = alpaka::test::unit::math::UnaryFunctorsComplex;
+using BinaryFunctorsComplex = alpaka::test::unit::math::BinaryFunctorsComplex;
+
+//! Caller of test template with additional lambda wrapping the functor
+//! @tparam TAcc Accelerator.
+//! @tparam TData Input data type..
+template<typename TAcc, typename TData>
+struct LambdaMathTestTemplate
+{
+    //! Run the test for one of the given functors
+    //! @tparam TFunctors Typelist of functors, each defined in Functor.hpp.
+    template<typename TFunctors>
+    void operator()() const
+    {
+        // To save on compiler memory and run time we test for only one of the functors and not all of them.
+        // The behavior should not differ between those as we already check all the functors without lambda.
+        // To avoid always using the first functor from the list, calculate index based on input data type sizes.
+        constexpr uint32_t index = (sizeof(TAcc) + sizeof(TData) + sizeof(TFunctors)) % std::tuple_size_v<TFunctors>;
+        using Functor = std::tuple_element_t<index, TFunctors>;
+        using ArgsItem = alpaka::test::unit::math::ArgsItem<TData, Functor::arity>;
+        auto wrappedFunctor
+            = [] ALPAKA_FN_HOST_ACC(ArgsItem const& arguments, TAcc const& acc) { return Functor{}(arguments, acc); };
+        auto testTemplate = TestTemplate<TAcc, Functor>{};
+        testTemplate.template operator()<TData>(wrappedFunctor);
+    }
+};
+
+TEMPLATE_LIST_TEST_CASE("mathOpsLambdaFloat", "[math] [operator]", TestAccs)
+{
+    using Acc = TestType;
+    auto testTemplate = LambdaMathTestTemplate<Acc, float>{};
+    testTemplate.template operator()<UnaryFunctorsReal>();
+    testTemplate.template operator()<BinaryFunctorsReal>();
+}
+
+TEMPLATE_LIST_TEST_CASE("mathOpsLambdaDouble", "[math] [operator]", TestAccs)
+{
+    using Acc = TestType;
+    auto testTemplate = LambdaMathTestTemplate<Acc, double>{};
+    testTemplate.template operator()<UnaryFunctorsReal>();
+    testTemplate.template operator()<BinaryFunctorsReal>();
+}
+
+TEMPLATE_LIST_TEST_CASE("mathOpsLambdaComplexFloat", "[math] [operator]", TestAccs)
+{
+    using Acc = TestType;
+    auto testTemplate = LambdaMathTestTemplate<Acc, alpaka::Complex<float>>{};
+    testTemplate.template operator()<UnaryFunctorsComplex>();
+    testTemplate.template operator()<BinaryFunctorsComplex>();
+}
+
+TEMPLATE_LIST_TEST_CASE("mathOpsLambdaComplexDouble", "[math] [operator]", TestAccs)
+{
+    using Acc = TestType;
+    auto testTemplate = LambdaMathTestTemplate<Acc, alpaka::Complex<double>>{};
+    testTemplate.template operator()<UnaryFunctorsComplex>();
+    testTemplate.template operator()<BinaryFunctorsComplex>();
+}
+#endif

